sdks/python/apache_beam/runners/portability/flink_runner_test.py

```diff
@@ -53,361 +53,380 @@
-        ]
-      finally:
-        rmtree(tmp_dir)
-
-    @classmethod
-    def get_runner(cls):
-      return portable_runner.PortableRunner()
-
-    @classmethod
-    def get_expansion_service(cls):
-      # TODO Move expansion address resides into PipelineOptions
-      return 'localhost:%s' % cls.expansion_port
-
-    def create_options(self):
-      options = super(FlinkRunnerTest, self).create_options()
-      options.view_as(
-          DebugOptions).experiments = ['beam_fn_api'] + extra_experiments
-      options._all_options['parallelism'] = 2
-      options.view_as(PortableOptions).environment_type = (
-          environment_type.upper())
-      if environment_config:
-        options.view_as(PortableOptions).environment_config = environment_config
-
-      if streaming:
-        options.view_as(StandardOptions).streaming = True
-      return options
-
-    # Can't read host files from within docker, read a "local" file there.
-    def test_read(self):
-      with self.create_pipeline() as p:
-        lines = p | beam.io.ReadFromText('/etc/profile')
-        assert_that(lines, lambda lines: len(lines) > 0)
-
-    def test_no_subtransform_composite(self):
-      raise unittest.SkipTest("BEAM-4781")
 
-    def test_external_transform(self):
+class FlinkRunnerTest(portable_runner_test.PortableRunnerTest):
+  _use_grpc = True
+  _use_subprocesses = True
+
+  conf_dir = None
+  expansion_port = None
+  flink_job_server_jar = None
+
+  def __init__(self, *args, **kwargs):
+    super(FlinkRunnerTest, self).__init__(*args, **kwargs)
+    self.environment_type = None
+    self.environment_config = None
+
+  @pytest.fixture(autouse=True)
+  def parse_options(self, request):
+    test_pipeline_options = (
+        request.config.option.test_pipeline_options
+        if request.config.option.test_pipeline_options else '')
+    parser = argparse.ArgumentParser(add_help=True)
+    parser.add_argument(
+        '--flink_job_server_jar',
+        help='Job server jar to submit jobs.',
+        action='store')
+    parser.add_argument(
+        '--environment_type',
+        default='LOOPBACK',
+        choices=['DOCKER', 'PROCESS', 'LOOPBACK'],
+        help='Set the environment type for running user code. DOCKER runs '
+        'user code in a container. PROCESS runs user code in '
+        'automatically started processes. LOOPBACK runs user code on '
+        'the same process that originally submitted the job.')
+    parser.add_argument(
+        '--environment_config',
+        help='Set environment configuration for running the user code.\n For '
+        'DOCKER: Url for the docker image.\n For PROCESS: json of the '
+        'form {"os": "<OS>", "arch": "<ARCHITECTURE>", "command": '
+        '"<process to execute>", "env":{"<Environment variables 1>": '
+        '"<ENV_VAL>"} }. All fields in the json are optional except '
+        'command.')
+    known_args, unknown_args = parser.parse_known_args(
+        test_pipeline_options.split())
```

Comment:
Could this be problematic if one of the option values contains a space, e.g. as part of a json string for the environment config?

Reply:
Good catch. I used `shlex.split(...)` instead.

Reply:
I also had to make many changes to preserve the correct string quoting. Tox in particular does something weird with quotes in its commands, so I had to double-escape the strings before passing to Tox and unescaping them after.